x-pack/legacy/plugins/lens/public/editor_frame_plugin/editor_frame/save.ts

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Action, EditorFrameState } from './state_management';
+import { Document } from '../../persistence/saved_object_store';
+
+export interface Props {
+  datasource: { getPersistableState: (state: unknown) => unknown };
+  dispatch: (value: Action) => void;
+  redirectTo: (path: string) => void;
+  state: EditorFrameState;
+  store: { save: (doc: Document) => Promise<{ id: string }> };
+  visualization: { getPersistableState: (state: unknown) => unknown };
+}
+
+export async function save({
+  datasource,
+  dispatch,
+  redirectTo,
+  state,
+  store,
+  visualization,
+}: Props) {
+  try {
+    dispatch({ type: 'SAVING' });
+
+    const doc = await store.save({
+      id: state.persistedId,
+      title: state.title,
+      type: 'lens',
+      visualizationType: state.visualization.activeId,
+      datasourceType: state.datasource.activeId,
+      state: {
+        datasource: datasource.getPersistableState(state.datasource.state),
+        visualization: visualization.getPersistableState(state.visualization.state),
+      },
+    });
+
+    if (doc.id !== state.persistedId) {
+      dispatch({ type: 'UPDATE_PERSISTED_ID', id: doc.id });
+      redirectTo(`/edit/${doc.id}`);
+    }
+  } finally {
+    dispatch({ type: 'SAVED' });
```

Comment:
I see that you are always emitting the `SAVED` action because it's not currently doing anything more than changing the button state, but like I commented earlier I think there should be three actions here. If we ever want to add more behavior on SAVE_ERROR we shouldn't need to also change the trigger.

Reply:
Yeah. I totally agree. I don't know what this should be called, though, so I settled with 'SAVED' since it's correct for now. I think maybe the current actions could be unified into one: `{ type: 'SAVING', isSaving: true | false }`. Dunno. We can tweak this kind of thing over time.